text/src/autogluon/text/automm/models/ft_transformer.py

```diff
@@ -0,0 +1,596 @@
+    def __init__(self, d_token: int, initialization: str) -> None:
+        """
+        Args:
+            d_token: the size of token
+            initialization: initialization policy for parameters. Must be one of
+                :code:`['uniform', 'normal']`. Let :code:`s = d ** -0.5`. Then, the
+                corresponding distributions are :code:`Uniform(-s, s)` and :code:`Normal(0, s)`. In
+                the paper [gorishniy2021revisiting], the 'uniform' initialization was
+                used.
+
+        References:
+            * [gorishniy2021revisiting] Yury Gorishniy, Ivan Rubachev, Valentin Khrulkov, Artem Babenko "Revisiting Deep Learning Models for Tabular Data", 2021
+        """
+        super().__init__()
+        initialization_ = _TokenInitialization.from_str(initialization)
+        self.weight = nn.Parameter(Tensor(d_token))
+        initialization_.apply(self.weight, d_token)
+
+    def expand(self, *leading_dimensions: int) -> Tensor:
+        """Expand (repeat) the underlying [CLS]-token to a tensor with the given leading dimensions.
+
+        A possible use case is building a batch of [CLS]-tokens. See `_CLSToken` for
+        examples of usage.
+
+        Note:
+            Under the hood, the `torch.Tensor.expand` method is applied to the
+            underlying :code:`weight` parameter, so gradients will be propagated as
+            expected.
+
+        Args:
+            leading_dimensions: the additional new dimensions
+
+        Returns:
+            tensor of the shape :code:`(*leading_dimensions, len(self.weight))`
+        """
+        if not leading_dimensions:
+            return self.weight
+        new_dims = (1,) * (len(leading_dimensions) - 1)
+        return self.weight.view(*new_dims, -1).expand(*leading_dimensions, -1)
+
+    def forward(self, x: Tensor) -> Tensor:
+        """Append self **to the end** of each item in the batch (see `_CLSToken`)."""
+        return torch.cat([x, self.expand(len(x), 1)], dim=1)
+
+
+class _TokenInitialization(enum.Enum):
+    UNIFORM = 'uniform'
+    NORMAL = 'normal'
+
+    @classmethod
+    def from_str(cls, initialization: str) -> '_TokenInitialization':
+        try:
+            return cls(initialization)
+        except ValueError:
+            valid_values = [x.value for x in _TokenInitialization]
+            raise ValueError(f'initialization must be one of {valid_values}')
+
+    def apply(self, x: Tensor, d: int) -> None:
+        d_sqrt_inv = 1 / math.sqrt(d)
+        if self == _TokenInitialization.UNIFORM:
+            # used in the paper "Revisiting Deep Learning Models for Tabular Data";
+            # is equivalent to `nn.init.kaiming_uniform_(x, a=math.sqrt(5))` (which is
+            # used by torch to initialize nn.Linear.weight, for example)
+            nn.init.uniform_(x, a=-d_sqrt_inv, b=d_sqrt_inv)
+        elif self == _TokenInitialization.NORMAL:
+            nn.init.normal_(x, std=d_sqrt_inv)
+
+
+class MultiheadAttention(nn.Module):
```

Comment:
Why can't we reuse [`torch.nn.modules.activation.MultiheadAttention`](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/activation.py#L887)?

Reply:
This is a good suggestion. @yiqings We may reuse https://pytorch.org/docs/1.11/generated/torch.nn.MultiheadAttention.html?highlight=multiheadattention#torch.nn.MultiheadAttention. We chose not to reuse TransformerEncoderLayer https://pytorch.org/docs/1.11/generated/torch.nn.TransformerEncoderLayer.html?highlight=transformerencoder#torch.nn.TransformerEncoderLayer because it does not support ReGLU. But MultiheadAttention will be good.